File: xor.nim
import os

# let defKey*: string = "HELLO"
var buf: array[4096, char]

let argv = commandLineParams()

if len(argv) <= 0:
  echo "Please enter key"
  quit(QuitFailure)

let key = argv[0]

var keyPos: int = 0

var bytesRead = readChars(stdin, buf, 0, len(buf))
while bytesRead > 0:
  var i:int = 0
  while i < bytesRead:
    buf[i] = char(int(buf[i]) xor int(key[i mod len(key)]))
    inc(i)
  discard writeChars(stdout, buf, 0, bytesRead)
  bytesRead = readChars(stdin, buf, 0, len(buf))

File: xor.c
#include <stdio.h>
#include <string.h>
#include <unistd.h>

char buf[4096];
const char *key;
int keyPos = 0;

char *defKey = "HELLO";

int main(int argc, char const *argv[]) {
  if (argc <= 1) {
    printf("Please enter key\n");
    return 1;
  } else
    key = argv[1];
  /* key = defKey; */

  int bytesRead = read(STDIN_FILENO, buf, 4096);
  while (bytesRead > 0) {
    for (int i = 0; i < bytesRead; i++) {
      buf[i] = buf[i] ^ key[i % strlen(key)];
    }
    write(STDOUT_FILENO, buf, bytesRead);
    bytesRead = read(STDIN_FILENO, buf, 4096);
  }

  return 0;
}
